File: tests/test_tether.cpp
#include <cmath>
#include <doctest/doctest.h>

#include "../include/manta/coupling/tether.hpp"
#include "../include/manta/core/craft.hpp"
#include "../include/manta/core/scene.hpp"
#include "../include/manta/core/world.hpp"
#include "../include/manta/parts/structure/point_mass.hpp"
#include "../include/manta/parts/coupling/tether_endpoint.hpp"
#include "test_helpers.hpp"

using namespace manta;
using namespace manta::geom;
using namespace manta::parts;
using namespace manta::coupling;

// ---- Tether dynamics unit tests ----

TEST_CASE("Tether: slack (d <= rest_length) → zero force") {
    Tether t(/*rest=*/10.0f, /*k=*/100.0f);
    auto F = t.force_on_self(
        Vec3<SceneFrame>{0, 0, 0},
        Vec3<SceneFrame>{5, 0, 0},  // separation 5 < 10
        Vec3<SceneFrame>::zero(),
        Vec3<SceneFrame>::zero());
    CHECK(test::approx_equal(F, Vec3<SceneFrame>::zero()));
}

TEST_CASE("Tether: at rest length exactly → zero force") {
    Tether t(10.0f, 100.0f);
    auto F = t.force_on_self(
        Vec3<SceneFrame>{0, 0, 0},
        Vec3<SceneFrame>{10, 0, 0},
        Vec3<SceneFrame>::zero(),
        Vec3<SceneFrame>::zero());
    CHECK(test::approx_equal(F, Vec3<SceneFrame>::zero()));
}

TEST_CASE("Tether: extended → Hooke's law toward sibling") {
    Tether t(/*rest=*/10.0f, /*k=*/100.0f);
    // d=12, extension=2, k=100 → |F|=200, direction +x (from self toward other)
    auto F = t.force_on_self(
        Vec3<SceneFrame>{0, 0, 0},
        Vec3<SceneFrame>{12, 0, 0},
        Vec3<SceneFrame>::zero(),
        Vec3<SceneFrame>::zero());
    CHECK(F.x() == doctest::Approx(200.0f));
    CHECK(F.y() == doctest::Approx(0.0f));
    CHECK(F.z() == doctest::Approx(0.0f));
}

TEST_CASE("Tether: equal-and-opposite forces on the two endpoints") {
    Tether t(10.0f, 100.0f);
    auto F_a = t.force_on_self(
        Vec3<SceneFrame>{0, 0, 0},
        Vec3<SceneFrame>{12, 0, 0},
        Vec3<SceneFrame>::zero(),
        Vec3<SceneFrame>::zero());
    auto F_b = t.force_on_self(
        Vec3<SceneFrame>{12, 0, 0},
        Vec3<SceneFrame>{0, 0, 0},
        Vec3<SceneFrame>::zero(),
        Vec3<SceneFrame>::zero());
    CHECK(test::approx_equal(F_a, Vec3<SceneFrame>::from_raw(-F_b.raw())));
}

TEST_CASE("Tether: damping resists separation when extended") {
    Tether t(10.0f, 100.0f, /*c=*/10.0f);
    // Stretched to 12 (extension 2), other moving away in +x at 1 m/s.
    // F_spring = 100 * 2 = 200; F_damp = 10 * 1 = 10; total = 210 (still toward +x).
    auto F = t.force_on_self(
        Vec3<SceneFrame>{0, 0, 0},
        Vec3<SceneFrame>{12, 0, 0},
        Vec3<SceneFrame>::zero(),
        Vec3<SceneFrame>{1, 0, 0});
    CHECK(F.x() == doctest::Approx(210.0f));
}

TEST_CASE("Tether: degenerate coincident points → zero force, no NaN") {
    Tether t(0.0f, 100.0f);
    auto F = t.force_on_self(
        Vec3<SceneFrame>::zero(),
        Vec3<SceneFrame>::zero(),
        Vec3<SceneFrame>::zero(),
        Vec3<SceneFrame>::zero());
    CHECK(test::approx_equal(F, Vec3<SceneFrame>::zero()));
}

// ---- TetherEndpoint integration tests ----

TEST_CASE("TetherEndpoint: stretched between two crafts → force pulls them together") {
    // Two 1-kg point-mass crafts, tether rest length 10, placed 12 apart
    // along +x, k=50 → |F| = 100 N on each. After 0.1s with 100 N on a 1kg
    // craft, |Δv| = 10 m/s. Crafts approach each other.
    Tether tether(/*rest=*/10.0f, /*k=*/50.0f, /*c=*/0.0f);

    World w;
    w.clock().set_dt(0.001f);
    auto& scene = w.create_scene();

    Craft c1("a"), c2("b");
    c1.root().add<PointMass>("body", 1.0f);
    c1.root().add<TetherEndpoint>("end", tether, /*is_a=*/true);
    c1.root().compute_params();
    c1.set_position(Vec3<SceneFrame>{0, 0, 0});
    scene.add_craft(c1);

    c2.root().add<PointMass>("body", 1.0f);
    c2.root().add<TetherEndpoint>("end", tether, /*is_a=*/false);
    c2.root().compute_params();
    c2.set_position(Vec3<SceneFrame>{12, 0, 0});
    scene.add_craft(c2);

    // Run a few ticks; verify both crafts have non-zero velocity toward each other.
    for (int i = 0; i < 50; ++i) w.update();

    auto v1 = c1.scene_to_craft().vel_linear();
    auto v2 = c2.scene_to_craft().vel_linear();

    // c1 starts at x=0, c2 at x=12. Tether pulls them together:
    // c1 should accelerate in +x, c2 in -x.
    CHECK(v1.x() > 0.0f);
    CHECK(v2.x() < 0.0f);
    // Equal masses + symmetric forces → equal-and-opposite velocities (within tol).
    CHECK(v1.x() == doctest::Approx(-v2.x()).epsilon(1e-3f));
}

TEST_CASE("TetherEndpoint: slack → no force, crafts drift independently") {
    Tether tether(20.0f, 100.0f);  // rest 20, separation will be 5

    World w;
    w.clock().set_dt(0.01f);
    auto& scene = w.create_scene();

    Craft c1("a"), c2("b");
    c1.root().add<PointMass>("body", 1.0f);
    c1.root().add<TetherEndpoint>("end", tether, true);
    c1.root().compute_params();
    c1.set_position   (Vec3<SceneFrame>{0, 0, 0});
    c1.set_vel_linear (Vec3<SceneFrame>{1, 0, 0});  // drift +x
    scene.add_craft(c1);

    c2.root().add<PointMass>("body", 1.0f);
    c2.root().add<TetherEndpoint>("end", tether, false);
    c2.root().compute_params();
    c2.set_position(Vec3<SceneFrame>{5, 0, 0});
    scene.add_craft(c2);

    for (int i = 0; i < 100; ++i) w.update();  // 1 s

    auto v1 = c1.scene_to_craft().vel_linear();
    auto v2 = c2.scene_to_craft().vel_linear();
    // Tether stays slack the whole time → c1 keeps its initial velocity, c2 stays at rest.
    CHECK(v1.x() == doctest::Approx(1.0f).epsilon(1e-4f));
    CHECK(test::approx_equal(v2, Vec3<SceneFrame>::zero(), 1e-4f));
}

TEST_CASE("TetherEndpoint: same-craft tether between two parts produces zero net force on craft") {
    // Two endpoints on the SAME craft, separated along x. Equal-and-opposite
    // forces should cancel at the craft level (no net translation). Internal
    // torque about CoM may exist, but the crafts CoM acceleration stays zero
    // because internal forces don't accelerate a closed system.
    Tether tether(1.0f, 100.0f);

    World w;
    w.clock().set_dt(0.01f);
    auto& scene = w.create_scene();

    Craft c("rigid");
    c.root().add<PointMass>("body", 1.0f);

    auto& a = c.root().add<TetherEndpoint>("a", tether, true);
    StaticLink<ParentFrame, PartFrame> tf_a{
        Vec3<ParentFrame>{ 2, 0, 0}, Ori<ParentFrame>::identity()};
    a.set_transform(tf_a);

    auto& b = c.root().add<TetherEndpoint>("b", tether, false);
    StaticLink<ParentFrame, PartFrame> tf_b{
        Vec3<ParentFrame>{-2, 0, 0}, Ori<ParentFrame>::identity()};
    b.set_transform(tf_b);

    c.root().compute_params();
    scene.add_craft(c);

    for (int i = 0; i < 50; ++i) w.update();

    // Internal force pair → no net translational motion of the craft origin.
    auto v = c.scene_to_craft().vel_linear();
    CHECK(test::approx_equal(v, Vec3<SceneFrame>::zero(), 1e-3f));
}

// ---- Scalar templating: TetherT<Jet> compiles & yields autodiff derivatives ----

#include <ceres/jet.h>

TEST_CASE("TetherT<Jet>: force_on_self differentiates through extension") {
    // Stretched tether: d=12, rest=10, k=100 → F_x = 200 (+x toward other).
    // Differentiate w.r.t. p_other_x: ∂F_x/∂p_other_x should equal stiffness
    // (k = 100) along the rhat direction. With both endpoints on the x-axis
    // and rhat = (+1,0,0), |F| = k*(d - rest); ∂|F|/∂p_other_x = k = 100.
    using Jet = ceres::Jet<double, 1>;
    coupling::TetherT<Jet> t(Jet(10.0), Jet(100.0), Jet(0.0));
    auto F = t.force_on_self(
        geom::Vec3<SceneFrame, Jet>{Jet(0), Jet(0), Jet(0)},
        geom::Vec3<SceneFrame, Jet>{Jet(12.0, 0), Jet(0), Jet(0)},
        geom::Vec3<SceneFrame, Jet>{Jet(0), Jet(0), Jet(0)},
        geom::Vec3<SceneFrame, Jet>{Jet(0), Jet(0), Jet(0)});
    CHECK(F.x().a       == doctest::Approx(200.0));
    CHECK(F.x().v(0)    == doctest::Approx(100.0));   // ∂F_x/∂p_other_x = k
    CHECK(F.y().a       == doctest::Approx(0.0));
    CHECK(F.z().a       == doctest::Approx(0.0));
}

TEST_CASE("TetherT<Jet>: slack region returns identically-zero Jet (no derivative leak)") {
    using Jet = ceres::Jet<double, 1>;
    coupling::TetherT<Jet> t(Jet(10.0), Jet(100.0));
    auto F = t.force_on_self(
        geom::Vec3<SceneFrame, Jet>{Jet(0), Jet(0), Jet(0)},
        geom::Vec3<SceneFrame, Jet>{Jet(5.0, 0), Jet(0), Jet(0)},  // d=5 < rest
        geom::Vec3<SceneFrame, Jet>{Jet(0), Jet(0), Jet(0)},
        geom::Vec3<SceneFrame, Jet>{Jet(0), Jet(0), Jet(0)});
    CHECK(F.x().a    == doctest::Approx(0.0));
    CHECK(F.x().v(0) == doctest::Approx(0.0));  // slack ⇒ derivative is zero
}

File: tests/test_tether.py
"""Tether coupling — multi-craft component dynamics."""

import numpy as np
import pytest

from manta import Craft, Sim, TargetNumpy, World
from manta.couplings import Tether
from manta.fields import GravityField
from manta.parts import Mass, TetherEndpoint


def _make_craft(name: str, mass: float = 1.0) -> Craft:
    c = Craft(name)
    c.add(Mass("body", mass=mass, moi=(0.1, 0.1, 0.1)))
    c.add(TetherEndpoint("hook"))
    return c


def test_at_rest_length_zero_force():
    """Two crafts exactly L_rest apart with zero velocity: no force, no
    motion."""
    L = 5.0
    a = _make_craft("a")
    b = _make_craft("b")
    w = World().add_field(GravityField.none())    # no gravity
    w.add_craft(a, position=(0, 0, 0))
    w.add_craft(b, position=(L, 0, 0))
    w.add_coupling(Tether(a, "hook", b, "hook",
                          stiffness=10.0, damping=1.0, rest_length=L))
    cw = TargetNumpy(Sim(w))

    for _ in range(500):
        cw.step(0.001)

    pa = np.array(cw.state["a"]["position"]).ravel()
    pb = np.array(cw.state["b"]["position"]).ravel()
    np.testing.assert_allclose(pa, np.zeros(3), atol=1e-9)
    np.testing.assert_allclose(pb, (L, 0, 0), atol=1e-9)


def test_stretched_spring_pulls_crafts_together():
    """Start crafts farther apart than rest length → the taut tether
    pulls them in; once inside rest length it goes slack, so the final
    separation is at most rest length (never pushed back out)."""
    L_rest = 5.0
    L_init = 7.0
    a = _make_craft("a")
    b = _make_craft("b")
    w = World().add_field(GravityField.none())
    w.add_craft(a, position=(0, 0, 0))
    w.add_craft(b, position=(L_init, 0, 0))
    w.add_coupling(Tether(a, "hook", b, "hook",
                          stiffness=10.0, damping=5.0, rest_length=L_rest))
    cw = TargetNumpy(Sim(w))

    for _ in range(2000):
        cw.step(0.001)

    pa = np.array(cw.state["a"]["position"]).ravel()
    pb = np.array(cw.state["b"]["position"]).ravel()
    dist = np.linalg.norm(pb - pa)
    # The pull happened, and the slack rope never pushed them back apart.
    assert dist < L_init - 0.5
    assert dist <= L_rest + 0.05


def test_slack_tether_exerts_no_force():
    """Start crafts closer than rest length → the tether is slack and
    exerts exactly zero force; nothing moves."""
    L_rest = 5.0
    L_init = 3.0
    a = _make_craft("a")
    b = _make_craft("b")
    w = World().add_field(GravityField.none())
    w.add_craft(a, position=(0, 0, 0))
    w.add_craft(b, position=(L_init, 0, 0))
    w.add_coupling(Tether(a, "hook", b, "hook",
                          stiffness=10.0, damping=5.0, rest_length=L_rest))
    cw = TargetNumpy(Sim(w))

    for _ in range(2000):
        cw.step(0.001)

    pa = np.array(cw.state["a"]["position"]).ravel()
    pb = np.array(cw.state["b"]["position"]).ravel()
    np.testing.assert_allclose(pa, np.zeros(3), atol=1e-12)
    np.testing.assert_allclose(pb, (L_init, 0, 0), atol=1e-12)


def test_damper_cannot_push_while_taut():
    """Barely taut with B closing fast: the raw spring+damper sum is
    strongly compressive, but a rope can't push — the tension clamp
    holds the force at ~zero instead of shoving A away from B."""
    a = _make_craft("a")
    b = _make_craft("b")
    w = World().add_field(GravityField.none())
    w.add_craft(a, position=(0, 0, 0))
    # Stretch = 0.1 m (spring +1 N), closing at 10 m/s (damper −50 N).
    w.add_craft(b, position=(5.1, 0, 0), velocity=(-10.0, 0, 0))
    w.add_coupling(Tether(a, "hook", b, "hook",
                          stiffness=10.0, damping=5.0, rest_length=5.0))
    cw = TargetNumpy(Sim(w))

    cw.step(0.001)

    va = np.array(cw.state["a"]["velocity"]).ravel()
    # A pushed away from B would move in −x. Old rigid-spring model:
    # Δv_a ≈ −49 N · 1 ms = −0.049 m/s. Clamped: essentially nothing.
    assert va[0] > -1e-6
    assert abs(va[0]) < 1e-3


def test_momentum_conservation_no_external_forces():
    """Two crafts coupled by a tether, no external forces → total
    momentum stays at its initial value forever."""
    a = _make_craft("a", mass=1.0)
    b = _make_craft("b", mass=2.0)
    w = World().add_field(GravityField.none())
    w.add_craft(a, position=(0, 0, 0), velocity=(1.0, 0, 0))
    w.add_craft(b, position=(7.0, 0, 0))
    w.add_coupling(Tether(a, "hook", b, "hook",
                          stiffness=100.0, damping=0.0, rest_length=5.0))
    cw = TargetNumpy(Sim(w))

    # Initial momentum
    va0 = np.array(cw.state["a"]["velocity"]).ravel()
    vb0 = np.array(cw.state["b"]["velocity"]).ravel()
    p0 = 1.0 * va0 + 2.0 * vb0   # m_a·v_a + m_b·v_b

    for _ in range(2000):
        cw.step(0.001)

    va = np.array(cw.state["a"]["velocity"]).ravel()
    vb = np.array(cw.state["b"]["velocity"]).ravel()
    p_final = 1.0 * va + 2.0 * vb
    np.testing.assert_allclose(p_final, p0, atol=1e-3)


def test_damping_dissipates_relative_motion():
    """With damping, oscillation amplitude shrinks over time."""
    a = _make_craft("a")
    b = _make_craft("b")
    w = World().add_field(GravityField.none())
    w.add_craft(a, position=(0, 0, 0))
    w.add_craft(b, position=(8.0, 0, 0))
    w.add_coupling(Tether(a, "hook", b, "hook",
                          stiffness=50.0, damping=10.0, rest_length=5.0))
    cw = TargetNumpy(Sim(w))

    # Run with damping. Elastic PE only accrues while taut (slack rope
    # stores nothing).
    energies = []
    for i in range(3000):
        cw.step(0.001)
        if i % 200 == 0:
            va = np.array(cw.state["a"]["velocity"]).ravel()
            vb = np.array(cw.state["b"]["velocity"]).ravel()
            pa = np.array(cw.state["a"]["position"]).ravel()
            pb = np.array(cw.state["b"]["position"]).ravel()
            ke = 0.5 * (np.dot(va, va) + np.dot(vb, vb))
            stretch = max(0.0, np.linalg.norm(pb - pa) - 5.0)
            pe = 0.5 * 50.0 * stretch ** 2
            energies.append(ke + pe)

    # Energy should be monotonically decreasing (with some tolerance for
    # discretization noise).
    assert energies[-1] < 0.1 * energies[0]


def test_offset_endpoint_produces_torque():
    """A tether attached at a body-frame offset transmits torque to the
    craft when stretched."""
    a = Craft("a")
    a.add(Mass("body", mass=1.0, moi=(0.05, 0.05, 0.05)))
    # Endpoint at +y offset → stretch along +x produces torque about z.
    a.add(TetherEndpoint("hook", mount_offset=(0.0, 0.5, 0.0)))

    b = Craft("b")
    b.add(Mass("body", mass=100.0, moi=(1.0, 1.0, 1.0)))   # heavy → stays put
    b.add(TetherEndpoint("hook"))

    w = World().add_field(GravityField.none())
    w.add_craft(a, position=(0, 0, 0))
    w.add_craft(b, position=(5.0, 0, 0))
    w.add_coupling(Tether(a, "hook", b, "hook",
                          stiffness=10.0, damping=0.0, rest_length=2.0))
    cw = TargetNumpy(Sim(w))

    cw.step(0.001)

    omega = np.array(cw.state["a"]["angular_velocity"]).ravel()
    # The tether pulls A's hook (at +y, +x relative to body center) toward B
    # (which is at +x in anchor). The force pulls A in +x direction, applied
    # at +y offset → torque about -z. Sign depends on convention.
    assert abs(omega[2]) > 1e-6
    # x and y components small (no torque about those axes by symmetry).
    assert abs(omega[0]) < 1e-9
    assert abs(omega[1]) < 1e-9


def test_tether_rejects_nested_endpoint():
    """An endpoint under a joint would get a silently wrong lever arm —
    the Tether constructor rejects it."""
    from manta.parts import RevoluteJoint

    a = Craft("a")
    a.add(Mass("body", mass=1.0, moi=(0.1, 0.1, 0.1)))
    j = a.add(RevoluteJoint("gim", axis=(0.0, 0.0, 1.0)))
    j.add(TetherEndpoint("hook"))
    b = _make_craft("b")
    import pytest
    with pytest.raises(ValueError, match="root"):
        Tether(a, "hook", b, "hook", stiffness=10.0)


def test_tether_endpoint_lookup_ignores_non_endpoints():
    """A same-named non-TetherEndpoint part must not satisfy the lookup."""
    a = Craft("a")
    a.add(Mass("hook", mass=1.0))      # decoy — not a TetherEndpoint
    b = _make_craft("b")
    with pytest.raises(ValueError, match="no TetherEndpoint"):
        Tether(a, "hook", b, "hook", stiffness=10.0)


@pytest.mark.parametrize(
    ("argument", "value"),
    (
        ("stiffness", -1.0),
        ("damping", -1.0),
        ("rest_length", -1.0),
        ("slack_smoothing", -1.0),
        ("stiffness", float("nan")),
        ("damping", float("inf")),
    ),
)
def test_tether_rejects_nonphysical_configuration(argument, value):
    a = _make_craft("a")
    b = _make_craft("b")
    kwargs = {"stiffness": 1.0, argument: value}
    with pytest.raises(ValueError, match=argument):
        Tether(a, "hook", b, "hook", **kwargs)
